User model

ferrine opened the pull request:
I started preparation for implementing GLM model and I think good base class is necessary for that purpose. Also refer to #1524 and #1517.

ferrine committed:
Started to write Base class for pymc3.models

ferrine committed:
mode `add_var` to public api

ferrine committed:
Added some docstrings

ferrine committed:
Added some docstrings

ferrine commented:
Do we really need compatible sklearn api? I think if well be a bit hard to make them friends
Are there any ideas of basic functionality of UserModel?

ferrine committed:
added getitem and fixed a typo

twiecki commented:
@ferrine I don't think it's a priority to keep sklearn API. Can just be a starting point but if it's not helpful there's no reason to be beholden to it.

ferrine committed:
added assertion check

ferrine committed:
added resolve var method

ferrine committed:
decided not to add resolve method

ferrine committed:
Added linear component

ferrine commented:
Please review LinearComponent API and base class usage example

ferrine committed:
Docs fix

ferrine committed:
patsy's intercept is inited properly now

ferrine commented:
Hmm, what's wrong with travis?

ferrine committed:
refactored code

ferrine committed:
updated docs

ferrine committed:
added possibility to init coefficients with random variables

twiecki commented on pymc3/models/linear.py:
```diff
@@ -0,0 +1,90 @@
+    init : dict - test_vals for coefficients
+    rvars : dict - random variables instead of creating new ones
+    """
+    def __init__(self, name, x, y, intercept=True, labels=None,
```
`name` is only required if you have multiple `LinearComponent`s, right? Not sure it should be a required arg then.

ferrine replied:
Agree

twiecki commented on pymc3/models/base.py:
```diff
@@ -0,0 +1,69 @@
+        """Shortcut to model"""
+        return modelcontext(None)
+
+    def new_var(self, name, dist, data=None, test_val=None):
```
I can see how this is a useful abstraction but it makes transcribing a PyMC3 model from the usual syntax to be a class a bit harder as it gives cognitive overhead. After all, the code below isn't doing so much that it warrants the indirection. Is there another way we can achieve that?

ferrine replied:
Unfortunately, there is no easier way to rename lots of variables in the model.

ferrine replied:
I also thought about a method like [this](https://github.com/pymc-devs/pymc3/pull/1525/files#diff-f85ae0776100e0962d318aaad0c5340fR88) for base class, but i'm not sure if it will be good for readability. Here it exists to avoid code duplication

ferrine committed:
added glm

ferrine committed:
refactored api, fixed formula init

ferrine commented:
travis was failing because of [that](https://travis-ci.org/pymc-devs/pymc3/jobs/175961733#L1020),  waiting for merge [this PR](https://github.com/Theano/Theano/pull/5232)

ferrine committed:
refactored linear model, extended acceptable types

ferrine committed:
moved useful matrix and labels creation to utils file

ferrine committed:
code style

ferrine committed:
removed redundant evaluation of shape

ferrine committed:
refactored resolver for constructing matrix and labels

ferrine committed:
changed error message

springcoil commented:
Are you ready to accept this @twiecki. I like the general idea.

springcoil reviewed (commented):
This looks good to me, I like some of those tests and the elegance of the abstraction. I find the naming lots of variables stuff to be a bit messy, but I can't think of an easier way and I think that naming lots of variables in Python is a bit of a pain.

ferrine commented:
I think that `any_to_tensor_and_labels` helper is not well tested yet, there is still some work left

ferrine committed:
changed signature of init

ferrine committed:
simplified utils any_to_tensor_and_labels code

ferrine committed:
tests for `any_to_tensor_and_labels`

ferrine commented:
Now the whole PR is tested. I can add some utils for visualisation for model variables. Do you have another ideas for improvements?

ferrine committed:
added docstring for `any_to_tensor_and_labels` util

ferrine committed:
forgot to document return type in `any_to_tensor_and_labels`

ferrine committed:
refactored code for dict

ferrine committed:
dict tests fix(do not check labels there)

ferrine committed:
Merge branch 'master' into user_model

ferrine committed:
added access to random vars of model

ferrine commented:
Hmm, what do you think, is api consistent with pymc3 ideology? 
I added properties for access to random variables but they are named a bit different from pm.Model. I expect that non random variables can be passed instead of RV. Is it really needed or I can rely on good user and make simplier api. 
What in general do you think about "how to get random variables defined inside the class"?

ferrine committed:
added a shortcut for all variables so there is a unified way to get them

ferrine committed:
added default priors for linear model

ferrine committed:
update docs for linear

ferrine committed:
refactored UserModel api, made it more similar to pm.Model class

ferrine committed:
Lots of refactoring, tests for base class, more plain api design

ferrine commented:
@twiecki, have a look please, I considered using nested model design finally. I think that fits well.

ferrine committed:
deleted unused module variable

ferrine committed:
fixed some typos in docstring

ferrine commented:
What about avoiding this tricky solution and just refactor `pm.Model` class so that we are able to add prefix to any variable if specified? That will give the same effect but in more straightforward way. Moreover I find not very difficult to wrap `__init__` phase in a context do handle model construction when creating a class

twiecki commented:
> What about avoiding this tricky solution and just refactor pm.Model class so that we are able to add prefix to any variable if specified? That will give the same effect but in more straightforward way. Moreover I find not very difficult to wrap **init** phase in a context do handle model construction when creating a class

I had the same thought :+1:.

ferrine commented:
@twiecki this idea will be real soon) 
In my opinion nested models should add variables to all models that are parents:

``` python
with m1:
    # adds to m1
    with m2:
        # adds to m1 and m2
        with m3:
            # adds to m1,2,3
```

Do you agree?

twiecki commented on pymc3/tests/test_models_linear.py:
```diff
@@ -0,0 +1,96 @@
+    def test_glm(self):
+        with Model() as model:
+            NAME = 'glm'
+            Glm(
```
And this works with supplying a `family` kwarg too?

twiecki commented:
I think this looks pretty good!

twiecki commented:
Although I'm not when someone would want to nest models?

ferrine commented:
There are several reasons

1. Incapsulate model creation in init phase.
    To archive that we need to run `__init__` in the context of instance(will be done in metaclass' `__call__`)
2. Catching init will help to rename(if name provided) arbitrary definitions in creation phase. It allows to avoid name overlap, just choose another name in init.
3. We will be able to logically separate a set of variables providing quick access to them. Might be useful for sampling not all variables later.

ferrine commented:
Moreover, to convert model to a class based model you will need to replace "with statement" with class' init method. A class should be a subclass of `pm.Model` then.

twiecki commented:
So would a model then look like:
`with pm.GLM(): ...` ?

twiecki commented:
Perhaps inheritance of `Model` isn't the right approach after all and composition is what's required. But that's what you currently have, no?

ferrine commented:
I still think about model nesting as the most elegant way to solve the problem of class based model definitions. Another way is to subclass from model to NestedModel class that can be used only in context, but I don't see reasons in that. 
I think that `with pm.Glm(...):` is good syntax as it denotes that we start with some use-case model, not empty one as in `with pm.Model():` and inside context `Glm(...)` has the same meaning that our linear model is in the context and thus should contribute in parent's model potentials and variables set. Names are needed to distinguish template models(e.g. Glm) definitions thus we need to add `name` argument to `pm.Model` as well as to all subclasses. To distinguish variables created within named/unnamed context we should add prefix to variables names. That is the reason of refactoring model class itself instead of hacking around.

ferrine committed:
Refactored pm.Model class, now it is ready for inheritance

ferrine committed:
Added documentation for Model class

ferrine commented:
@twiecki, most of the work is done, I added documentation for "How to subclass" in pm.Model. A good example of subclassing is [Linear Component](https://github.com/pymc-devs/pymc3/pull/1525/files#diff-f85ae0776100e0962d318aaad0c5340fR15) or [Glm](https://github.com/pymc-devs/pymc3/pull/1525/files#diff-f85ae0776100e0962d318aaad0c5340fR90)
In order to simplify reading, here it is
```python
class CustomModel(Model):
    # 1) override init
    def __init__(self, mean=0, sd=1, name='', model=None):
        # 2) call super's init first, passing model and name to it
        # name will be prefix for all variables here
        # if no name specified for model there will be no prefix
        super(CustomModel, self).__init__(name, model)
        # now you are in the context of instance,
        # `modelcontext` will return self
        # you can define variables in several ways
        # note, that all variables will get model's name prefix

        # 3) you can create variables with Var method
        self.Var('v1', Normal.dist(mu=mean, sd=sd))
        # this will create variable named like '{prefix_}v1'
        # and assign attribute 'v1' to instance
        # created variable can be accessed with self.v1 or self['v1']

        # 4) this syntax will also work as we are in the context
        # of instance itself, names are given as usual
        Normal('v2', mu=mean, sd=sd)

        # something more complex is allowed too
        Normal('v3', mu=mean, sd=HalfCauchy('sd', beta=10, testval=1.))

        # Deterministic variables can be used in usual way
        Deterministic('v3_sq', self.v3 ** 2)
        # Potentials too
        Potential('p1', tt.constant(1))

# After defining a class CustomModel you can use it in several ways

# I:
#   state the model within a context
with Model() as model:
    CustomModel()
    # arbitrary actions

# II:
#   use new class as entering point in context
with CustomModel() as model:
    Normal('new_normal_var', mu=1, sd=0)

# III:
#   just get model instance with all that was defined in it
model = CustomModel()

# IV:
#   use many custom models within one context
with Model() as model:
    CustomModel(mean=1, name='first')
    CustomModel(mean=2, name='second')
```

ferrine committed:
Small typo in docstring

ferrine committed:
nested contains for treedict (needed for add_random_variable)

ferrine committed:
More accurate duplicate implementation of treedict/treelist

ferrine committed:
refactored treedict/treelist

ferrine committed:
changed `__imul__` of treelist

ferrine commented:
Done. Need review

twiecki commented:
I'm still not sure why we need two ways of defining a model. Why is `with pm.Model(): Glm()...` worse than `with pm.Glm(): ...`? In which case would the second one be preferable? If it didn't exist, we also wouldn't have to deal with nesting and all the complexity this adds here.

ferrine commented:
@twiecki, first method is not worse than second. All that comes from inheritance, I think that's even better. All cases are valid, so you can choose what like more. For me second is convenient if I have only one model `Glm` and I simply run
```python
with pm.Glm(...):
    trace = pm.sample(1000)
```
First can be more convenient if I have some more models
```python
with pm.Model() as model:
    pm.Glm(..., name='one')
    pm.Glm(..., name='two')
    trace = pm.sample(1000)
```

ferrine commented:
Some comments about nesting. Here is `__new__` of base model
```python
def __new__(cls, *args, **kwargs):
    # resolves the parent instance
    instance = object.__new__(cls)
    # if user provides parent model set it as parent
    # new instance will be nested within it
    if kwargs.get('model') is not None:    
        instance.parent = kwargs.get('model')
    # user provides no parent model, check context stack
    elif cls.get_contexts():
        instance.parent = cls.get_contexts()[-1]    # great, set it as parent
    else:
        instance.parent = None    # we are out of the context, no parent model
    instance.model = instance
    return instance
```
I also have a metaclass to change context for init
```python
class InitContextMeta(type):
    """Metaclass that executes `__init__` of instance in it's context"""
    def __call__(cls, *args, **kwargs):
        # __new__ was described above
        instance = cls.__new__(cls, *args, **kwargs)
        # run init in the context of instance
        # that's how we can control names of variables in `__init__`
        with instance:  # appends context
            instance.__init__(*args, **kwargs)
        return instance
```
There is no magic except `treelist` and  `treedict` classes that provide functionality for passing created variables to parent model instance if needed(not None) as well as checking that we create duplicate var. I've tested them a lot so that I'm sure they will work as I expect they should

twiecki commented:
I think at this point we should invite some thoughts from others. CC @fonnesbeck @ColCarroll maybe?

My opinion is this: I appreciate the effort that went into this and flexibility this provides. It's well written and tested code. But if we abandon the inheritance from `pm.Model` and go back to encapsulation we get two things of value:
1. There is only one way to use this new API, consistent with what we had before ("There should be one-- and preferably only one --obvious way to do it.")
2. Less code complexity.

This PR currently adds features we don't know we need and to which there are easy alternatives: inheritance and model nesting. We already do enough magic with the model context.

Instead, I'd rather merge an earlier version, we see how far we get with it and if we do run into issues we can go back to the implementation that's already here.

ferrine commented:
> There is only one way to use this new API, consistent with what we had before ("There should be one-- and preferably only one --obvious way to do it.")

From that point of view I'd rather change a few lines of code in `__new__` instead of rolling back to old api idea.
```python
else:
    if not type(instance) is Model:
        raise TypeError("No context on context stack")
    instance.parent = None    # we are out of the context, no parent model
```
I think that previous api I suggested was error prone due to usage mistakes. It was enforcing user to use new api to create a model instead of common one(I'm about creating class based model and it's init). That is additional cost.

fonnesbeck commented:
While I appreciate an object-oriented approach as a developer, I'm less enthusiastic about it as a modeler. I really don't want our users to have to write classes and subclassing existing classes in order to build models. I'm not opposed to a secondary API for very advanced users wanting to do more complicated things (assuming our current API does not accommodate them), but I think our current API is preferable for users who are not computer scientists. The overreaching aim of the PyMC syntax is to have users spend as little time as possible thinking about the mechanics of Bayesian modeling as possible, allowing them to focus on the scientific problem they are trying to model.

Is there a good case that shows what this API does that our current one does not? I would require some convincing examples for me to vote for changing course.

ferrine commented:
> The overreaching aim of the PyMC syntax is to have users spend as little time as possible thinking about the mechanics of Bayesian modeling as possible, allowing them to focus on the scientific problem they are trying to model.

That's exactly what I want to archive.

## Problems 
1. Current API does not support reusing the model later. You need to copy and paste model from the previous project. I solve this problem with encapsulating this routine in a class. So that for some givens you get a model. I hope contributors will help with creating a collection of frequently used models or create there own repos for domain specific models. Related work is in progress(#1517), some issues are coming(#1547), this can be a part of advanced tutorial(#1488).
2. I want to allow user to create his model with building blocks. Now his is done for `pm.glm` only. Moreover, making functions for creating these building blocks is error prone stuff (see #1513). For making this building blocks for user we must be sure that nothing fails unexpectedly. In my case it was a sudden duplicate variable, so I copied, pasted and refactored that module to continue experimenting. 
Functional approach is good when you have "one model - one task". You just wrap model in a function and that is done(see this [module](https://github.com/quantopian/pyfolio/blob/master/pyfolio/bayesian.py)). 
Problems appear when you have one model - many tasks, like linear model. You can't say it solves exactly one task, because it can be used standalone as well as for modelling heteroscedacity. But it is true not only for glm. At present you don't have useful building blocks nor api to create one. [Lasagne](https://github.com/Lasagne/Lasagne) has that option via inheritance and that's really great.
3. Not only routine on modelling spends final user's time.
Some models require nontrivial data manipulations and pymc3 variable creations that assume deep python/theano/pymc3 knowledge just for one glm call.
For example, [here](https://github.com/pymc-devs/pymc3/issues/1512#issuecomment-262100123) I write this stuff in the class as a general solution, so my further experiments will be done in one line(that's what we need). If not reused, these solutions are just a waste of time that could be spent on experiments and problem solving, especially when it is something simple. Moreover ad-hoc solutions make code unreadable, hard to understand and error prone. 
If I am a user and want to run, for instance, a panel regression with fixed effects in many specifications, to chose the best, I understand that I have to write lot's of python code, not related to the problem. If I find ready implementation in pymc3 it will save a lot of my time, if I will be able to write the solution once in the class, it will be the second best.

Thus we have three main problems.
1. No code reusage
2. No chance to have building blocks
3. Lots of python is required to experiment with some simple things(no encapsulation)

## Changes
That's why I decided to refactor `Model`. Every single change is needed. They provide
1. **Context awareness** 
It is done by `treedict`, `treelist` and `__new__`. User will get exactly what he needs when he uses a class based model. 
2. **Possibility to create mode than one model**
It is archived with `name_for`, `name_of` methods. Now you will never get in trouble with overlapped names as in(#1513).
3. **Easy contributing**
With metaclass, all what is needed to create a reusable model is to change "with statement" with class definition and optionally open a PR.

## Solutions
Together, changes solve these problems. 
1. Code reusage is done via classes. It will be not a trouble for final user.
2. Building blocks are ok and will not break
3. Subclassing when needed will encapsulate dirty python stuff and save time for problem solving. Having a module `pm.models` you will allow contributors to share their experience and use-case models. Or else any can create repo with domain specific models.

I do not want to force users to write a class for every single task. I want to save their time with predefined models and make it easy to create ones for contributors.

twiecki commented:
I definitely agree on the motivation here.

When dropping the `with Glm()` syntax, are you saying we would still need the metaclass and treedict and treelist?

ferrine commented:
Yes, we need them for clear work of all the system.
 I thought a little more and found that we can still use a subclass `with newclass: ...` inside the context if we check only "out of context instance" in the "fix" above. To remove that behaviour some tricks should be implemented. I think that it is cheaper (in terms of [code complexity] - [one way to use]) to leave it without changes.

twiecki commented:
What are your thoughts on model-specific inference and model-specific plots? How would those be integrated here?

ferrine commented:
What do you mean under model specific inference? Treat the environment as given and see only on the our little part?
I appreciate model specific plots. They can be implemented directly in subclasses or in base model with option to override.

ferrine committed:
added `root` property and `isroot` indicator for base model

ferrine committed:
protect `parent` and `model` attributes from violation

ferrine commented:
Though I don't think model specific plots are a part of this PR.

ferrine commented:
Not in WIP. It's done.

ferrine committed:
travis' python2 did not fail on bad syntax(maybe it's too new), fixed

ferrine committed:
decided not to use functools wrapper

 Unfortunately functools wrapper fails
 when decorating built-in methods so I
 need to fix that improper behaviour.
 Some bad but needed tricks were implemented

ferrine commented:
Travis plays bad tricks, definitely. I relied on his tests but in my new projects environment (python 2.7.12) model crushed.

ferrine committed:
Added models package to setup script

ColCarroll commented on pymc3/model.py:
```diff
@@ -168,17 +169,213 @@ def logpt(self):
+            return list.__contains__(self, item)
+
+    def __setitem__(self, key, value):
+        raise NotImplementedError('Not able to determine '
```
Can exception message be more specific?  Something like "cannot set item value for `treelist` (tried to set {key} = {value})"?

ferrine replied:
I cannot set value at all. It inserts a value by some key, but in parent list there can be something else. I'm not sure if we need that, but this will probably cause unexpected behaviour.

ColCarroll replied:
Ah, I just meant the exception message could be more clear!

ColCarroll commented on pymc3/models/utils.py:
```diff
@@ -0,0 +1,92 @@
+import theano.tensor as tt
+
+
+def any_to_tensor_and_labels(x, labels=None):
```
any chance for breaking up this logic a bit?

ColCarroll replied:
(just saw awesome test suite, but still comment still holds :) )

ColCarroll replied:
It looks like splitting out `_input_to_array`, `_verify_labels`, then you can rewrite this as

```
x = _input_to_array(x)
labels = _verify_labels(labels)
assert_shape_labels(x.shape, labels)
return x, labels
```

ferrine replied:
labels can come from x, so they go together

ColCarroll commented on pymc3/models/utils.py:
```diff
@@ -0,0 +1,92 @@
+    -------
+    (x, labels) - tensor and labels for it's columns
+    """
+    def assert_shape_labels(s, l):
```
looks like this is only used once

ColCarroll commented on pymc3/models/utils.py:
```diff
@@ -0,0 +1,92 @@
+            raise ValueError('Cannot use scalars')
+        elif x.ndim == 1:
+            x = x[:, None]
+    # else: trust input
```
would love an explicit
```
else:
    pass
```
here

ColCarroll replied:
(and below)

ColCarroll reviewed (approved):
Looks really good!  I'll have to look at the examples more, but seems fine to merge to me.  I left some nitpicks that are probably better off as followup PRs.

ColCarroll commented:
Also, there are a few uses of "it's" (="it is") in the docs that should change to "its" (possessive).

ferrine committed:
Refactor utils

ferrine committed:
Fix some typos in pm.model

ferrine commented:
BTW, [gelato](https://github.com/ferrine/gelato) is comming. PyMC3 becomes really powerful

twiecki commented:
Alright, I'll go ahead and merge this into the `3.1` branch. We'll merge it into master once we shipped 3.0. Thanks for the great contribution @ferrine.

twiecki commented:
Just checking out Gelato, looks really interesting!
